splice_machine/src/main/java/com/splicemachine/derby/stream/function/NLJInnerJoinFunction.java

```diff
@@ -122,6 +172,105 @@ public OperationContext getOperationContext() {
 
     @Override
     public LocatedRow getLeftLocatedRow() {
-        return leftRow;
+        return new LocatedRow(leftRowLocation, leftRow);
+    }
+
+    private void init(Iterator<LocatedRow> from) throws StandardException {
+        checkInit();
+        SConfiguration configuration= EngineDriver.driver().getConfiguration();
+        batchSize = configuration.getNestedLoopJoinBatchSize();
+        nLeftRows = 0;
+        leftSideIterator = from;
+        executorService = Executors.newFixedThreadPool(batchSize);
+        completionService = new ExecutorCompletionService<>(executorService);
+
+        initOperationContexts();
+        loadBatch();
+    }
+
+    private void initOperationContexts() throws StandardException {
+        try {
+            operationContextList = new ArrayList<>(batchSize);
+            for (int i = 0; i < batchSize; ++i) {
+                operationContextList.add(operationContext.getClone());
+            }
+        }
+        catch (Exception e) {
+            throw Exceptions.parseException(e);
+        }
+    }
+
+    private void loadBatch() throws StandardException {
+
+        try {
+            while (nLeftRows < batchSize) {
+                if (!leftSideIterator.hasNext())
+                    break;
+                nLeftRows++;
+                LocatedRow locatedRow = leftSideIterator.next();
+                completionService.submit(new GetRightScanIterator(operationContextList.remove(0), locatedRow));
+            }
+            if (nLeftRows > 0) {
+                Future<Pair<OperationContext, Iterator<LocatedRow>>> future = completionService.take();
+                Pair<OperationContext, Iterator<LocatedRow>> result = future.get();
+                currentOperationContext = result.getFirst();
+                rightSideNLJIterator = result.getSecond();
+                leftRow = currentOperationContext.getOperation().getLeftOperation().getCurrentRow();
+                leftRowLocation = currentOperationContext.getOperation().getLeftOperation().getCurrentRowLocation();
+                operationContext.getOperation().getLeftOperation().setCurrentLocatedRow(getLeftLocatedRow());
+                nLeftRows--;
+            }
+        }
+        catch (Exception e) {
+            throw Exceptions.parseException(e);
+        }
+    }
+
+    private void setRightSideCurrentLocatedRow(SpliceOperation from, SpliceOperation to) {
+        try {
+            try {
+                ExecRow row = from.getCurrentRow();
+                RowLocation location = from.getCurrentRowLocation();
+                to.setCurrentRow(row);
+                to.setCurrentRowLocation(location);
+            }
+            catch (Exception e) {
+                return;
+            }
+            List<SpliceOperation> fromSubOperations = from.getSubOperations();
```

Comment:
Jun mentioned the recursive copy is needed for RowID, a comment here would be good.